File: [iOS] HEALTH 1ST/maps.ios-master/NMapSampleObjc/NMapSampleObjc/MapViewController.m
//
//  MapViewController.m
//  NMapSampleObjc
//
//  Created by Naver on 2016. 11. 10..
//  Copyright © 2016년 Naver. All rights reserved.
//

#import "MapViewController.h"
#import "NMapViewResources.h"

@interface MapViewController ()

@property (nonatomic, strong) NMapView * mapView;
@property (weak, nonatomic) IBOutlet UIStepper *levelStepper;

@end

@implementation MapViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.mapView = [[NMapView alloc] initWithFrame:self.view.frame];
    self.navigationController.navigationBar.translucent = NO;
    
    // set the delegate for map view
    [self.mapView setDelegate:self];
    
    // set the application api key for Open MapViewer Library
    [self.mapView setClientId:@"YOUR CLIENT ID"];
 
    self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    [self.view insertSubview:self.mapView atIndex:0];

    // set min and max level for range in the UIStepper
    [self initLevelStepper:self.mapView.minZoomLevel maxValue:self.mapView.maxZoomLevel initialValue:11];
    [self.view bringSubviewToFront:self.levelStepper];
    
    [self.mapView setBuiltInAppControl:YES];
}

- (void) viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    [self.mapView setFrame:self.view.bounds];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.mapView viewWillAppear];
}

- (void)viewDidAppear:(BOOL)animated {
    [self.mapView viewDidAppear];
    [super viewDidAppear:animated];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.mapView viewWillDisappear];
}

- (void)viewDidDisappear:(BOOL)animated {
    [self.mapView viewDidDisappear];
    [super viewDidDisappear:animated];
}

#pragma mark NMapViewDelegate Method

- (void) onMapView:(NMapView *)mapView initHandler:(NMapError *)error {
    if (error == nil) { // success
        // set map center and level
        [self.mapView setMapCenter:NGeoPointMake(126.978371, 37.5666091) atLevel:11];
        // set for retina display
        [self.mapView setMapEnlarged:YES mapHD:YES];
        // set default map mode
        [self.mapView setMapViewMode:NMapViewModeVector];
    } else { // fail
        NSLog(@"onMapView:initHandler: %@", [error description]);
    }
}

#pragma mark NMapPOIdataOverlayDelegate

- (UIImage *) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay imageForOverlayItem:(NMapPOIitem *)poiItem selected:(BOOL)selected {
    return [NMapViewResources imageWithType:poiItem.poiFlagType selected:selected];
}

- (CGPoint) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay anchorPointWithType:(NMapPOIflagType)poiFlagTyp{
    return [NMapViewResources anchorPoint:poiFlagTyp];
}

- (UIImage*) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay imageForCalloutOverlayItem:(NMapPOIitem *)poiItem
           constraintSize:(CGSize)constraintSize selected:(BOOL)selected
imageForCalloutRightAccessory:(UIImage *)imageForCalloutRightAccessory
          calloutPosition:(CGPoint *)calloutPosition calloutHitRect:(CGRect *)calloutHitRect {
    return nil;
}

- (CGPoint) onMapOverlay:(NMapPOIdataOverlay *)poiDataOverlay calloutOffsetWithType:(NMapPOIflagType)poiFlagType {
    return CGPointZero;
}

#pragma mark MapViewController

- (IBAction)showOverlayLayers:(UIBarButtonItem *)sender {

    UIAlertController* alertController = [UIAlertController alertControllerWithTitle:@"Overlay Layers"
                                                                   message:nil
                                                                 preferredStyle:UIAlertControllerStyleActionSheet];

    UIAlertAction* trafficAction = [UIAlertAction actionWithTitle:[NSString stringWithFormat:@"Traffic layer is %@", (self.mapView.mapViewTrafficMode ? @"On" : @"Off")]
                                                           style:UIAlertActionStyleDefault
                                                         handler:^(UIAlertAction * action) {
                                                             self.mapView.mapViewTrafficMode = !self.mapView.mapViewTrafficMode;
                                                         }];
    UIAlertAction* bicycleAction = [UIAlertAction actionWithTitle:[NSString stringWithFormat:@"Bicycle layer is %@", (self.mapView.mapViewBicycleMode ? @"On" : @"Off")]
                                                            style:UIAlertActionStyleDefault
                                                          handler:^(UIAlertAction * action) {
                                                              self.mapView.mapViewBicycleMode = !self.mapView.mapViewBicycleMode;
                                                          }];
    UIAlertAction* alphaAction = [UIAlertAction actionWithTitle:[NSString stringWithFormat:@"Alpha layer is %@", (self.mapView.mapViewAlphaLayerMode ? @"On" : @"Off")]
                                                          style:UIAlertActionStyleDefault
                                                        handler:^(UIAlertAction * action) {
                                                            self.mapView.mapViewAlphaLayerMode = !self.mapView.mapViewAlphaLayerMode;
                                                            
//                                                            지도 위 반투명 레이어에 색을 지정할 때에는 다음 메서드를 사용한다
//                                                            [self.mapView setMapViewAlphaLayerMode:!self.mapView.mapViewAlphaLayerMode
//                                                                                         withColor:[UIColor colorWithRed:0.5 green:1.0 blue:0.5 alpha:0.9]];
                                                        }];

    [alertController addAction:trafficAction];
    [alertController addAction:bicycleAction];
    [alertController addAction:alphaAction];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alertController animated:YES completion:nil];
}

- (IBAction)changeMapMode:(UISegmentedControl *)sender {
    
    switch ([sender selectedSegmentIndex]) {
        case 0:
            [self.mapView setMapViewMode:NMapViewModeVector];
            break;
        case 1:
            [self.mapView setMapViewMode:NMapViewModeSatellite];
            break;
        case 2:
            [self.mapView setMapViewMode:NMapViewModeHybrid];
            break;
        default:
            [self.mapView setMapViewMode:NMapViewModeVector];
            break;
    }
}

- (void) initLevelStepper:(NSInteger)minValue maxValue:(NSInteger)max initialValue:(NSInteger) init {
    self.levelStepper.minimumValue = minValue;
    self.levelStepper.maximumValue = max;
    self.levelStepper.stepValue = 1;
    self.levelStepper.value = init;
    
}
- (IBAction)levelStepperValeChanged:(UIStepper *)sender {
    [self.mapView setZoomLevel:sender.value];
}

@end

File: [iOS] HEALTH 1ST/maps.ios-master/NMapSampleSwift/NMapSampleSwift/MapViewController.swift
//
//  MapViewController.swift
//  NMapSampleSwift
//
//  Created by Junggyun Ahn on 2016. 11. 9..
//  Copyright © 2016년 Naver. All rights reserved.
//

import UIKit

class MapViewController: UIViewController, NMapViewDelegate, NMapPOIdataOverlayDelegate {

    var mapView: NMapView?

    @IBOutlet weak var levelStepper: UIStepper!

    override func viewDidLoad() {
        super.viewDidLoad()

        mapView = NMapView(frame: self.view.bounds)

        if let mapView = mapView {

            // set the delegate for map view
            mapView.delegate = self

            // set the application api key for Open MapViewer Library
            mapView.setClientId("YOUR CLIENT ID")

            mapView.autoresizingMask = [.flexibleWidth, .flexibleHeight]

            view.addSubview(mapView)
            
            // Zoom 용 UIStepper 셋팅.
            initLevelStepper(mapView.minZoomLevel(), maxValue:mapView.maxZoomLevel(), initialValue:11)
            view.bringSubview(toFront: levelStepper)

            mapView.setBuiltInAppControl(true)
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        
        mapView?.didReceiveMemoryWarning()
    }

    // MARK: - NMapViewDelegate

    open func onMapView(_ mapView: NMapView!, initHandler error: NMapError!) {
        if (error == nil) { // success
            // set map center and level
            mapView.setMapCenter(NGeoPoint(longitude:126.978371, latitude:37.5666091), atLevel:11)
            // set for retina display
            mapView.setMapEnlarged(true, mapHD: true)
            // set map mode : vector/satelite/hybrid
            mapView.mapViewMode = .vector
        } else { // fail
            print("onMapView:initHandler: \(error.description)")
        }
    }

    // MARK: - NMapPOIdataOverlayDelegate

    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, imageForOverlayItem poiItem: NMapPOIitem!, selected: Bool) -> UIImage! {
        return NMapViewResources.imageWithType(poiItem.poiFlagType, selected: selected)
    }

    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, anchorPointWithType poiFlagType: NMapPOIflagType) -> CGPoint {
        return NMapViewResources.anchorPoint(withType: poiFlagType)
    }

    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, calloutOffsetWithType poiFlagType: NMapPOIflagType) -> CGPoint {
        return CGPoint(x: 0, y: 0)
    }

    open func onMapOverlay(_ poiDataOverlay: NMapPOIdataOverlay!, imageForCalloutOverlayItem poiItem: NMapPOIitem!, constraintSize: CGSize, selected: Bool, imageForCalloutRightAccessory: UIImage!, calloutPosition: UnsafeMutablePointer<CGPoint>!, calloutHit calloutHitRect: UnsafeMutablePointer<CGRect>!) -> UIImage! {
        return nil
    }

    // MARK: - Layer Button
    @IBAction func layerButtonAction(_ sender: UIBarButtonItem) {

        let alertController = UIAlertController(title: "Layers", message: nil, preferredStyle: .actionSheet)

        if let map = mapView {
            // Action Sheet 생성

            let trafficAction = UIAlertAction(title: "Traffic layer is " + (map.mapViewTrafficMode ? "On" : "Off"), style: .default, handler: { (action) -> Void in
                print("Traffic layer Selected...")
                map.mapViewTrafficMode = !map.mapViewTrafficMode
            })

            let bicycleAction = UIAlertAction(title: "Bicycle layer is " + (map.mapViewBicycleMode ? "On" : "Off"), style: .default, handler: { (action) -> Void in
                print("Traffic layer Selected...")
                map.mapViewBicycleMode = !map.mapViewBicycleMode
            })
            
            let alphaAction = UIAlertAction(title: "Alpha layer is " + (map.mapViewAlphaLayerMode ? "On" : "Off"), style: .default, handler: { (action) -> Void in
                print("Alpha layer Selected...")
                map.mapViewAlphaLayerMode = !map.mapViewAlphaLayerMode
                
//                지도 위 반투명 레이어에 색을 지정할 때에는 다음 메서드를 사용한다
//                map.setMapViewAlphaLayerMode(!map.mapViewAlphaLayerMode, with: UIColor(red: 0.5, green: 1.0, blue: 0.5, alpha: 0.9))
            })

            alertController.addAction(trafficAction)
            alertController.addAction(bicycleAction)
            alertController.addAction(alphaAction)
        }

        alertController.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))

        self.present(alertController, animated: true, completion: nil)
    }

    // MARK: - Map Mode Segmented Control

    @IBAction func modeChanged(_ sender: UISegmentedControl) {
        switch sender.selectedSegmentIndex {
        case 0:
            mapView?.mapViewMode = .vector
        case 1:
            mapView?.mapViewMode = .satellite
        case 2:
            mapView?.mapViewMode = .hybrid
        default:
            mapView?.mapViewMode = .vector
        }
    }

    // MARK: - Level Stepper

    func initLevelStepper(_ minValue: Int32, maxValue: Int32, initialValue: Int32) {
        levelStepper.minimumValue = Double(minValue)
        levelStepper.maximumValue = Double(maxValue)
        levelStepper.stepValue = 1
        levelStepper.value = Double(initialValue)
    }

    @IBAction func levelStepperValeChanged(_ sender: UIStepper) {
        mapView?.setZoomLevel(Int32(sender.value))
    }
}
